craft_platforms/_build.py

```diff
@@ -0,0 +1,55 @@
+# This file is part of craft-platforms.
+#
+# Copyright 2025 Canonical Ltd.
+#
+# This program is free software: you can redistribute it and/or modify it
+# under the terms of the GNU Lesser General Public License version 3, as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful, but WITHOUT
+# ANY WARRANTY; without even the implied warranties of MERCHANTABILITY,
+# SATISFACTORY QUALITY, or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License along
+# with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""General build planner for any app."""
+
+from typing import Any, Callable, Dict, Iterable
+
+from craft_platforms import charm, rock, snap
+from craft_platforms._buildinfo import BuildInfo
+from craft_platforms._platforms import get_platforms_build_plan
+
+_APP_SPECIFIC_PLANNERS: Dict[str, Callable[..., Iterable[BuildInfo]]] = {
+    "charmcraft": charm.get_platforms_charm_build_plan,
+    "rockcraft": rock.get_rock_build_plan,
+    "snapcraft": snap.get_platforms_snap_build_plan,
+}
+
+
+def get_build_plan(
+    app: str,
+    *,
+    project_data: Dict[str, Any],
+) -> Iterable[BuildInfo]:
+    """Get a build plan for a given application.
+
+    :param app: The name of the application (e.g. snapcraft, charmcraft, rockcraft)
+    :param project_data: The raw dictionary of the project's YAML file. Normally this
+        is what's output from ``yaml.safe_load()``.
+    :returns: An iterable containing each possible BuildInfo for this file.
+
+    This function is an abstraction layer over the general build planners, taking
+    the application's name and its raw data and returning an exhaustive build plan
+    for the given project for that application. This allows craft-platforms to be used
+    in a forward-compatible manner as it adds more logic around selecting build
+    planners or special behaviour for more apps.
+    """
+    planner = _APP_SPECIFIC_PLANNERS.get(app, get_platforms_build_plan)
```

Comment:
* Wdyt about raising a friendly error for unsupported *craft apps?

Reply:
I think that's fine, but then what do you think about returning the default build planner anything besides snapcraft, rockcraft, and charmcraft?

The scenario I am thinking about is:
1. Snarfcraft is created and uses the default build planner
2. Launchpad adds support to build snarfcraft and uses the default build planner
3. Snarfcraft adds a special rule and thus adds a new Snarfcraft build planner to craft-platforms
4. Launchpad can't just update the `craft-platforms` dep, they have to change which build planner they import and use

Reply:
That's the point of this feature. The snarfcraft owner is responsible for adding their custom entrypoint to `_APP_SPECIFIC_PLANNERS` so launchpad *can* just update `craft-platforms` as long as they're calling `craft_platforms.get_build_plan("snarfcraft", project_data)`

Reply:
Making an executive decision to release as-is since we can break this with 0.7 if I'm misunderstanding your concerns.

Reply:
I will dm you, because I want to talk about a specific example, rather than snarfcraft.